wicketd/src/update_tracker.rs

```diff
@@ -1844,6 +1950,76 @@ impl UpdateContext {
         })
     }
 
+    async fn interrogate_rot_bootloader(
+        &self,
+        rot_bootloader: &[ArtifactIdData],
+    ) -> Result<StepResult<RotInterrogation>, UpdateTerminalError> {
+        // We already validated at repo-upload time there is at least one RoT
+        // artifact available and that all available RoT artifacts are the same
+        // version, so we can unwrap the first artifact here and assume its
+        // version matches any subsequent artifacts.
+        // TODO this needs to be fixed for multi version to work!
+        let available_artifacts_version = rot_bootloader
+            .get(0)
+            .expect("no RoT artifacts available")
+            .id
+            .version
+            .clone();
+
+        // Read the caboose of the currently running version (always 0)
+        // When updating from older stage0 we may not have a caboose so an error here
+        // need not be fatal
+        // TODO make this fatal at some point
+        let caboose = self
+            .mgs_client
+            .sp_component_caboose_get(
+                self.sp.type_,
+                self.sp.slot,
+                SpComponent::STAGE0.const_as_str(),
+                0,
+            )
+            .await
+            .map(|v| v.into_inner())
+            .ok();
+
+        let available_artifacts = rot_bootloader.to_vec();
+        let make_result = |active_version| RotInterrogation {
+            // We always update slot 1
+            slot_to_update: 1,
+            available_artifacts,
+            available_artifacts_version,
+            sp: self.sp,
+            active_version,
+        };
+
+        match caboose {
+            Some(c) => {
+                let message = format!(
+                    "RoT bootloader version {} (git commit {})",
+                    c.version, c.git_commit
+                );
+
+                match c.version.parse::<SemverVersion>() {
+                    Ok(version) => StepSuccess::new(make_result(Some(version)))
+                        .with_message(message)
+                        .into(),
+                    Err(err) => StepWarning::new(
+                        make_result(None),
+                        format!(
+                            "{message} (failed to parse RoT bootloader version: {err})"
+                        ),
+                    )
+                    .into(),
+                }
+            }
+            None => StepSuccess::new(make_result(None))
+                .with_message(
+                    "Stage0 has no caboose, proceeding with update anyway",
```

Comment:
There are few enough stage0 versions released that we could include an FWID to known version map here until we've purged them from existence.

Reply:
I can think of 7 off the top of my head (staging/dev and prod/rel for each of gimlet psc rot plus bart). Anything that isn't one of those should probably just be updated manually via faux-mgs